### ISSUE TITLE
Justification for task-relevant batchnorm usage

### ISSUE THREAD

**Issue by tuanpham96** [Laborieux-Axel/SynapticMetaplasticityBNN#2]
This might be a silly question, but is there a justification for saving and using task-dependent batchnorm states? 

I understand that practically saving it is "cheaper" as it scales only linearly. 

1. But with the pMNIST task or MNIST-USPS task (`scenario = task`), and for example the network is deployed "in the wild" somehow, how would the network "know" which set of batchnorm states to use when an image is given? Would it just iterate over all sets of batchnorm states and return the output with the highest probability or confidence? 

2. For the SI/EWC solutions, as I understand, these also require (re)loading task-relevant saved batchnorm states, did you also run experiments without loading in these states? Are these solutions supposed to figure out task boundaries that might replace the need to remember/store these states?

3. For the CIFAR dataset (`scenario = domain`), there was a comparison between using task-dependent batchnorm and task-independent instance norm in the supplemental. The latter still achieved good solution but not as good as batchnorm. Were there also experiments on of instance norm on the pMNIST model or MNIST-USPS as well?

I apologize in advance if I misinterpre or misread something in the paper. 
Thank you!

**Comment by Laborieux-Axel** [comment]
Hi,

Thanks for your interest in the work! 

The difference between having and not having learnable batchnorm parameters is that it (understandably) increases the number of tasks that can be learned sequentially when we allow task specific batchnorm params.
Now the problem is that because the batchnorm params are not binarized (nor binarizable in a straightforward way), it is not clear how to consolidate them with our approach, so we focused instead on the binarized weights and allowed for task dependent batchnorm parameters for all the methods we compared. Note however that for the stream learning setting we explicitly removed learnable BN params. When different tasks are learned sequentially, the right BN params to be used could be inferred from a context or attention mechanism (for instance based on the statistics of the activations). This is an interesting further development for this work.

I unfortunately did not have time to test all the combinations of settings with using instance norm, but it is possible to test them using this repo.
